### PR TITLE
wivrn: init at 0.15

### PR DESCRIPTION
**Pull request by PassiveLemon**
## Description of changes
Adds WiVRn
https://github.com/NixOS/nixpkgs/pull/278126#issuecomment-1975239007

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ckiee** [comment]
Continuation of https://github.com/NixOS/nixpkgs/pull/293058

**Review by ckiee — CHANGES_REQUESTED** [review]
Looks like all the previous reviews have been more-or-less addressed so it's just up to someone testing it with a headset. 

Putting in a _request changes_ to prevent accidental merges until it's tested. I'll try to do it whenever I get my headset out.

**Comment by PassiveLemon** [comment]
As far as I can tell, WiVRn works fine, it's Steam and Proton that are causing problems: https://github.com/ValveSoftware/steam-runtime/issues/575

**Comment by PassiveLemon** [comment]
How would I use a package that isn't merged into nixpkgs yet but is in a different PR? I would like to install `monado-vulkan-layers` from https://github.com/NixOS/nixpkgs/pull/301044#issuecomment-2033106270 in the module

**Comment by Scrumplex** [comment]
> How would I use a package that isn't merged into nixpkgs yet but is in a different PR? I would like to install `monado-vulkan-layers` from [#301044 (comment)](https://github.com/NixOS/nixpkgs/pull/301044#issuecomment-2033106270) in the module

You could cherry-pick the commit and then drop it during a rebase once the other PR gets merged.

File: pkgs/by-name/wi/wivrn/package.nix
Comment (technobaboo) [reply pkgs/by-name/wi/wivrn/package.nix:72]: agreed, making this change made it work for me

**Comment by technobaboo** [comment]
Everything works just fine for me with the lynx r1 and nixOS (besides video encoding settings issues) so yeah, ship it!

(it might be nice to put those video encoding settings in the config though if possible)

**Comment by PassiveLemon** [comment]
> Everything works just fine for me with the lynx r1 and nixOS (besides video encoding settings issues) so yeah, ship it!
> 
> (it might be nice to put those video encoding settings in the config though if possible)

I don't see a reason to put those in the config. They will be used if they are enabled in WiVRn's config so all we need to do is just make sure they can be built

**Comment by technobaboo** [comment]
is there anything left to do before this can be merged? it's proven to work and all

**Comment by PassiveLemon** [comment]
Not that I know of. Just waiting for reviewers to do their thing

**Review by ckiee — APPROVED** [review]

**Comment by ckiee** [comment]
Approved with the assumption the recent pushes are just rebasing. I'll have a look at merging tomorrow (:

**Comment by PassiveLemon** [comment]
> Approved with the assumption the recent pushes are just rebasing. I'll have a look at merging tomorrow (:

Yep, there was a conflict in the 24.05 release notes file so I fixed that.

**Review by Pandapip1 — APPROVED** [review]

**Comment by awtterpip** [comment]
any timeline for this getting merged?

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (Scrumplex) [inline nixos/doc/manual/release-notes/rl-2405.section.md:178]: looks like you accidentally duplicated these changelog entries

File: pkgs/by-name/wi/wivrn/package.nix
Comment (Scrumplex) [inline pkgs/by-name/wi/wivrn/package.nix:8]:
```suggestion
, cudaPackages ? { }
```

File: pkgs/by-name/wi/wivrn/package.nix
Comment (Scrumplex) [inline pkgs/by-name/wi/wivrn/package.nix:35]:
```suggestion
, cudaSupport ? config.cudaSupport
}:
```

File: pkgs/by-name/wi/wivrn/package.nix
Comment (Scrumplex) [inline pkgs/by-name/wi/wivrn/package.nix:67]:
```suggestion
```

File: pkgs/by-name/wi/wivrn/package.nix
Comment (Scrumplex) [inline pkgs/by-name/wi/wivrn/package.nix:91]:
```suggestion
  ] ++ lib.optionals cudaSupport [
    cudaPackages.cuda_cudart
  ];
```

**Review by Scrumplex — APPROVED** [review]
This would make it easier for people to use this package who don't have CUDA hardware

File: pkgs/by-name/wi/wivrn/package.nix
Comment (Pandapip1) [inline pkgs/by-name/wi/wivrn/package.nix:97]: Could you add a comment stating why NVENC is disabled?

File: pkgs/by-name/wi/wivrn/package.nix
Comment (Pandapip1) [inline pkgs/by-name/wi/wivrn/package.nix:94]: I'd would like it if the client build was enabled and put in a seperate output, but I wouldn't consider this a blocker.

File: pkgs/by-name/wi/wivrn/package.nix
Comment (Pandapip1) [inline pkgs/by-name/wi/wivrn/package.nix:62]: I think @Scrumplex missed this reference to `cudaPackages`:

```suggestion
  nativeBuildInputs = [
    cmake
    git
    pkg-config
    python3
  ] ++ lib.optionals cudaSupport [
    cudaPackages.cuda_nvcc
  ];
```

File: pkgs/by-name/wi/wivrn/package.nix
Comment (Pandapip1) [inline pkgs/by-name/wi/wivrn/package.nix:54]: Any reason why you can't just depend o monado, or use the monado derivation's `src` attribute?

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
I second @Scrumplex's request to make CUDA optional.

File: pkgs/by-name/wi/wivrn/package.nix
Comment (ckiee) [reply pkgs/by-name/wi/wivrn/package.nix:94]: Indeed, not a blocker. The PC test client build seems to be an internal artifact and is absolutely not relevant for the server part, and is therefore out of scope for this PR.

File: pkgs/by-name/wi/wivrn/package.nix
Comment (ckiee) [reply pkgs/by-name/wi/wivrn/package.nix:54]: WiVRn vendors a monado source and builds it in its build system, we override `FETCHCONTENT_SOURCE_DIR_MONADO`.

Comment (Pandapip1) [reply pkgs/by-name/wi/wivrn/package.nix:54]: That makes sense. Thanks for clarifying!

**Comment by PassiveLemon** [comment]
I had forgotten to push changes I made a bit ago. Could you review the updated commits? I believe Cuda isn't required to build now

**Comment by Goxore** [comment]
Tried building and running the wivrn-server binary, and it says that the version is 0.0.0.0, causing version mismatch error on the quest. Just me?

**Comment by PassiveLemon** [comment]
Are you sure you installed the right version APK to your Quest? I've always seen v0.0.0 as well but only encountered version mismatch errors when the server and Quest were different versions.

**Comment by Goxore** [comment]
I am using the 0.14.1 prebuilt release APK, which I assume to be compatible.
https://github.com/Meumeu/WiVRn/releases/tag/v0.14.1

**Comment by Pandapip1** [comment]
Result of `nixpkgs-review pr 299830` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>wivrn</li>
  </ul>
</details>

**Comment by PassiveLemon** [comment]
> I am using the 0.14.1 prebuilt release APK, which I assume to be compatible. https://github.com/Meumeu/WiVRn/releases/tag/v0.14.1

Not quite sure what your issue is then, I can't really help with that

**Comment by ckiee** [comment]
make sure to uninstall first before `adb install`ing, it can be picky sometimes

**Review by ckiee — APPROVED** [review]
Code LGTM, going for merge.
#### :white_check_mark: build

Result of `nixpkgs-review pr 299830` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>wivrn</li>
  </ul>
</details>

#### checklist
- [X] server runs w/o nixos module (just avahi cfg)
- [X] server runs w nixos module
- [ ] needs testing w android client

**Comment by PassiveLemon** [comment]
This push fixes Cuda not being found and makes cudaSupport optional for NVENC support

File: pkgs/by-name/wi/wivrn/package.nix
Comment (xytovl) [inline pkgs/by-name/wi/wivrn/package.nix:136]: There are a few "automatic" options that should be set manually:
`WIVRN_USE_SYSTEM_BOOST` requires boost 1.75 for server, 1.84 for Wireshark plugin
`WIVRN_USE_SYSTEMD` for pretty hostname and to use a safer codepath when using monado
`WIVRN_USE_PIPEWIRE` / `WIVRN_USE_PULSEAUDIO` for audio backends

**Comment by PassiveLemon** [comment]
Xytovl pointed out to me in Discord that the URLs for the patches are based on the master branch and that I should change it to the version tag. My problem is, I don't know how to reference the wivrn version attribute in the monadoSrc derivation. Using ${finalAttrs.version} will return the version specified for Monado. How should I go about this?

**Comment by Pandapip1** [comment]
> Xytovl pointed out to me in Discord that the URLs for the patches are based on the master branch and that I should change it to the version tag. My problem is, I don't know how to reference the wivrn version attribute in the monadoSrc derivation. Using ${finalAttrs.version} will return the version specified for Monado. How should I go about this?

Use `(monado.src + "/path/to/file")`.

**Review by Pandapip1 — APPROVED** [review]
Diff LGTM

**Comment by PassiveLemon** [comment]
uh what

**Comment by PassiveLemon** [comment]
I swear I just did a simple rebase

**Comment by drupol** [comment]
It looks like you accidentally mass-pinged a bunch of people, which are now subscribed
and getting notifications for everything in this pull request. Unfortunately, they
cannot be automatically unsubscribed from the issue (removing review request does not
unsubscribe), therefore development cannot continue in this pull request anymore.

Please create a new pull request, link back to this one and ping the
people actually involved in here over there. For the next time, remember to set your PR to draft status before rebasing. In draft status, you can preview the list of maintainers that are about to be requested for review, which allows you to sidestep this issue.

In order to avoid this in the future, there are instructions for how to properly
rebase between branches in our [contribution guidelines](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging).
Setting your pull request to draft prior to rebasing is strongly recommended.
In draft status, you can preview the list of people that are about to be requested
for review, which allows you to sidestep this issue.
This is not a bulletproof method, though, as OfBorg still does review requests even on draft PRs.

**Pull request closed without merging**
